Update hocuspocus data exchange on collaborative documents

brunopagno opened the pull request:
# Ticket
https://community.openproject.org/projects/communicator-stream/work_packages/66328/activity

# What are you trying to accomplish?
Counterpart of https://github.com/opf/op-blocknote-hocuspocus/pull/2

Setting up API based collaboration using a YJS provider.

- Generate an OAuth token on demand when a new user opens a document
- Generate the actual OAuth application on demand if there isn't one for hocuspocus
- Authenticate requests with the YJS provider passing the OAuth token, document name and document id (they are properly validated on the provider)
- Define/Update GET and PUT endpoints for communication with the YJS provider
- Cover with the appropriate tests

brunopagno committed:
send both document id and name

brunopagno committed:
store yjs binary + new document API endpoint to send binary data

brunopagno committed:
remove unnecessary threadstore configuration

brunopagno committed:
define both get and put endpoints for binary data communication with hocuspocus server

brunopagno committed:
fix existing tests

brunopagno committed:
use base64 approach instead of binary

brunopagno committed:
blocknote editor use ydoc when collaboration is disabled

brunopagno committed:
fix lint errors and fix unused parameter

akabiru commented on modules/documents/db/migrate/20250923064241_add_y_doc_blob_to_documents.rb:
```diff
@@ -0,0 +1,35 @@
+
+class AddYDocBlobToDocuments < ActiveRecord::Migration[8.0]
+  def change
+    add_column :documents, :content_binary, :text
```
🟢 Curious why the data type ended up as "text", instead of "bytea"? I think you've mentioned it before but I missed it

brunopagno replied:
Forgot to update here before. So there are a few reasons, but the main one is that the API is going to work with base64 strings representing the binary. So it felt unnecessary to transform it for the database.

Now the follow up question is why the API is using text instead of binary streams. The reason is that when sending binary data you basically send only the binary stream. So the hocuspocus server would need to make at least two requests to update the document on the server. One with the binary content and another with some form of plaintext content as we want to persist data that is searcheable and that can be used on the API.

It also makes everything much simpler as the API currently uses json for everything. And querying and serving the data is also simpler. So it felt like using text was the better way.

akabiru reviewed (approved):
Nice one ✨  🟠 Only thing I noticed that's missing is API Docs, we should probably add that soon?

brunopagno committed:
use oauth token instead of custom JWT for auth

NobodysNightmare commented on modules/documents/app/services/documents/oauth/ensure_application_service.rb:
```diff
@@ -0,0 +1,70 @@
+      private
+
+      def find_or_create_application
+        existing = Doorkeeper::Application.find_by(name: APPLICATION_NAME)
```
I'd avoid searching/matching applications by name, because the name is supposed to be human-readable.

We could even think about localizing the name one day.

The machine readable identifier is the Client ID, which doorkeeper stores as `uid`. The `uid` is not required to be gibberish. It's perfectly allowed to be something like `hocuspocus`, so we can hardcode it as constant (like the name).

brunopagno replied:
Absolutely lovely :bow:

NobodysNightmare commented on modules/documents/app/services/documents/oauth/generate_token_service.rb:
```diff
@@ -0,0 +1,66 @@
+        application.access_tokens.create(
+          resource_owner_id: @user.id,
+          scopes: "api_v3",
+          expires_in: 24.hours.to_i
```
I'd question whether such a long expiration time is necessary. The default we configure is two hours.

Without having read the rest of the code: Our implementation should also work well if the expiration time was one minute. Or in other words: The token expiring in the middle of the session should be a case that we consider and that we can deal with.

I am pointing this out, because 24 hours looks like a value chosen to ensure that tokens never expire while we use them.

```suggestion
```

brunopagno replied:
Yeah, I agree that this is the way to go, refreshing tokens and all, I just didn't want to do everything in one go. I'd like to merge with 24h, though, and tackle this separately.

I am not sure how to measure what's the real risk of going live with this solution, though, in case we don't implement a refresh flow in time.

NobodysNightmare replied:
Let's just make sure to have at least a work package reminding us to do this.

I am not too panicked about the 24 hours, because also our default of two hours is pretty long to be honest. I'd prefer us to bring this down to something in the realm of "a few minutes" at some point.

brunopagno replied:
Let such work package be - https://community.openproject.org/wp/68460

NobodysNightmare commented on modules/documents/app/services/documents/oauth/generate_token_service.rb:
```diff
@@ -0,0 +1,66 @@
+        application_result = EnsureApplicationService.new.call
+        return application_result unless application_result.success?
+
+        application = application_result.result
```
I am wondering whether we should make the so-called `ServiceResultRefinements` part of the core implementation of `ServiceResult`.

For example we could be using `bind` here as far as I can tell:

```ruby
EnsureApplicationService.new.call.bind do |application|
  token = create_access_token(application)

  if token.persisted?
    ServiceResult.success(result: token)
  else
    ServiceResult.failure(errors: token.errors)
  end
end
```

CC @mereghost

brunopagno committed:
updated documents on documents API

brunopagno committed:
adjust hocuspocus admin settings + urls + oauth application

brunopagno committed:
add missing specs

akabiru commented on frontend/src/react/OpBlockNoteContainer.tsx:
```diff
@@ -64,99 +60,99 @@ const detectTheme = ():OpColorMode => { return window.OpenProject.theme.detectOp
-  if(collaborationEnabled) {
-    const resolveUsers = async (userIds:string[]) => {
-      return users.filter((user) => userIds.includes(user.id));
+    editorParams = {
+      schema,
+      collaboration: {
+        provider: null,
+        fragment: doc.getXmlFragment('document-store'),
+        user: {
+          name: activeUser.username,
+          color: '#333333',
+        },
+      },
```
🟢 nit: looks like we can DRY `editorParams` building? A fn could provide additional context as well

akabiru commented on frontend/src/react/OpBlockNoteContainer.tsx:
```diff
@@ -64,99 +60,99 @@ const detectTheme = ():OpColorMode => { return window.OpenProject.theme.detectOp
-      url: hocuspocusUrl,
-      name: documentId,
-      token: hocuspocusAccessToken,
+      url: `${hocuspocusUrl}?document_id=${documentId}&openproject_base_path=${encodeURIComponent(openProjectUrl)}`,
```
🟢 `URL` object builder might be more elegant and less brittle than string based concatenation.

Ex. https://github.com/opf/openproject/blob/b1ba50be99cf64f71af2e1244454bcb72f4e63a0/frontend/src/stimulus/controllers/dynamic/work-packages/activities-tab/lazy-page.controller.ts#L99-L107

brunopagno replied:
True that. Looks way nicer!

akabiru commented on frontend/src/react/OpBlockNoteContainer.tsx:
```diff
@@ -64,99 +60,99 @@ const detectTheme = ():OpColorMode => { return window.OpenProject.theme.detectOp
+      if (collaborationEnabled && hocuspocusProvider) {
         hocuspocusProvider.destroy();
+      } else {
+        doc.off('update', updateInput);
```
what does this mean? 😅 Inline comment perhaps?

brunopagno replied:
It's the opposite of `doc.on('event_name'...);`

But fair enough :+1:

akabiru commented on modules/documents/app/services/documents/oauth/ensure_application_service.rb:
```diff
@@ -0,0 +1,72 @@
+
+module Documents
+  module OAuth
+    class EnsureApplicationService < BaseServices::BaseCallable
```
🟢 Perhaps a brief comment on the flow would be useful for others?

akabiru reviewed (approved):
Updates look good to me 👍🏾 Some few remarks, feel free to ignore

brunopagno committed:
adjust details for merge

- fix building the URL to hook up the Hocuspocus provider on Blocknote
- remove from documentation a field that should not be available yet on the API (should be released in future versions)
- add a feature flag check on the new documents PATCH endpoint, if it's not enabled we don't take requests yet
